### PR TITLE
[Merged by Bors] - feat: Port `Algebra.Hom.Ring`

### PR DESCRIPTION
**Pull request by winstonyin**
mathlib3 SHA: cf9386b56953fb40904843af98b7a80757bbe7f9

- [x] depends on: #949

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#949
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#949~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by winstonyin** [comment]
There are two more lint errors remaining, pertaining to some `comp` lemmas. Not sure how to fix them without writing long constructors explicitly.

File: Mathlib/Algebra/Hom/Ring.lean [hunk 1]
```diff
@@ -0,0 +1,788 @@
...
+#align function.injective.is_domain Function.Injective.isDomain
+
+namespace AddMonoidHom
+
+variable [CommRing α] [IsDomain α] [CommRing β] (f : β →+ α)
+
+/-- Make a ring homomorphism from an additive group homomorphism from a commutative ring to an
+integral domain that commutes with self multiplication, assumes that two is nonzero and `1` is sent
+to `1`. -/
+def mkRingHomOfMulSelfOfTwoNeZero (h : ∀ x, f (x * x) = f x * f x) (h_two : (2 : α) ≠ 0)
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Ring.lean:753]: This naming surprised me a bit; does the `_of_` naming scheme not apply for data?

Comment (winstonyin) [reply Mathlib/Algebra/Hom/Ring.lean:753]: I'm not aware of any exceptions to data naming scheme when `of` is involved, and I would slightly prefer not creating such an exception

Comment (semorrison) [reply Mathlib/Algebra/Hom/Ring.lean:753]: Asked on zulip: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/naming.20conventions/near/315558410

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**